### ISSUE TITLE
Bring back the .NET support

### ISSUE THREAD

**Issue by radare** [radare/radare2#9055]
long time ago we had the msil disassembler in core, but nothing was done in the PE side.

looks like yara have this parser, so we can reuse it. but will be good to handle PE as a fat fileformat to handle DOS/WIN*/.NET

https://github.com/VirusTotal/yara/blob/master/libyara/modules/dotnet.c

**Comment by Maijin** [comment]
Dup https://github.com/radare/radare2/issues/662

**Comment by radare** [comment]
Its not dupe. This can be implemented without making it fat and points to a working implementation that we can reuse

**Comment by JohnPeng47** [comment]
Is there no one working on this? If not would like to give it a shot, looking to get a deeper understanding of PE/.NET formats. Is there current support for the MSIL disassembly in radare?

**Comment by JohnPeng47** [comment]
Is essentially what you are asking highlighted in the following GSOC proposal?

    1st term: Refactoring RBin to support FAT and multibin
    2nd term: Support of FAT binaries (Win32 native + .NET) with new RBin, basic one
    Final: Also should be working with listing symbols from both parts of the binary (e.g. .NET and native code), as well as other metadata. And show this metadata in rabin2 output as well.

**Comment by radare** [comment]
See in extras. I moved the msil disasm plugin there. But no fatpe support yet and disasm plugin is broken

**Comment by JohnPeng47** [comment]
When you say fatpe, are you referring to mixed assemblies that contain native and managed code?

**Comment by radare** [comment]
Yes, PE binaris have multiple sub binaries inside. One of 1x86-16 for DOS, a 32bit for windows, and msil for .net. The idea is that pe must work like fatmach0 and be able to select which sub bin to open

**Comment by JohnPeng47** [comment]
After comparing the execution of rabin2 -A on a fat mach-o, single bin mach-o and  a pe file, I found that the fat mach-o will pass the check_bytes check, leading to a call to r_bin_file_xtr_load_bytes, which eventually determines the # of archs present in the fat binary. 
```
r_list_foreach (bin->binxtrs, it, xtr) 
	if (xtr && xtr->check_bytes (buf_bytes, sz)) { 
        .....
         binfile = r_bin_file_xtr_load_bytes (bin, xtr,
               fname, buf_bytes, sz, file_sz,
               baseaddr, loadaddr, xtr_idx,
               fd, bin->rawstr);

```
Looking at the check_bytes:
```
static bool check_bytes(const ut8* buf, ut64 sz) {
	return buf && sz > 3 && !memcmp (buf, "dyld", 4);
}
```
My initial assumption of the check_bytes function was that it is used to determine whether or not the bin file is a fat mach-o file, and if it is, then proceed with extracting the sub binaries. Is this what it's used for?

**Comment by radare** [comment]
yeah, ideally we should redesign the whole rbin api to make it more consistent. things has started to be done, but enotime for everything, and dont expect that to change from one day to other.

checkbytes is to check if the file is of a known type. subbinaries in fatmach0 are others mach0s, and in this csae its also checking for dyldcaches

**Comment by radare** [comment]
cool!

**Comment by JohnPeng47** [comment]
Wut exactly would be the output of the .Net extractor plugin? Because unlike fatmacho which is e just a container for multiple sub-binaries, mixed assembly .Net does have such a clear separation. If we pull apart the mixed assembly binary into one with with purely unmanaged code and one with pure managed code, neither of them will execute properly since the managed/unmanaged components depend on each other.

**Comment by radare** [comment]
maybe i would do sub-sets of the PE, this is, removing the higher layers

DOS bin should have no windows or .net code. for example

**Comment by JohnPeng47** [comment]
Ok so from one mixed PE file, then we should have 3 separate, properly formed PE files:
1. DOS bin (without native and .NET code)
2. Native code (without .NET with DOS?)
3. .NET code (with DOS and Native?)

**Comment by Maijin** [comment]
I guess if we want them to be workable we probably need to keep the DOS for native and the native and DOS part for .NET?

**Comment by JohnPeng47** [comment]
I was thinking the same thing where we can extract the Native and DOS code into its own bin, and then leave the original mixed assembly .NET untouched.

**Comment by radare** [comment]
Yes for yhe dos we can do a tiny bin with rabin2 -C directly. But nullifying some pointers and truncating the file should be the same

**Comment by JohnPeng47** [comment]
Is the configure-plugins script still used to build the plugins?

**Comment by radare** [comment]
Yes. Its executed by configure so you dont have to run it unless you want to configure which plugins you want to build and which as static or dynamic

**Comment by Maijin** [comment]
rebase and PR

**Comment by JohnPeng47** [comment]
new PR: https://github.com/radare/radare2/pull/10835

**Comment by JohnPeng47** [comment]
Affirmative
